### ISSUE TITLE
Follow XDG Base Directory Specification

### ISSUE THREAD

**Issue by ssokolow** [sagiegurari/cargo-make#105]
#### Problem Description

Putting dotfiles and dotdirs like `.cargo-make` directily inside `$HOME` is deprecated legacy behaviour.

The [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) provides guidelines for where things should go.

Rather than `~/.cargo-make/config.toml`, claim a single name inside `$XDG_CONFIG_HOME`. If that environment variable is not set, fallback to `$HOME/.config`.

Rather than `~/.cargo-make/cache.toml`, claim a single name inside `$XDG_CACHE_HOME`. If that environment variable is not set, fallback to `$HOME/.cache`.

**TL;DR:**
* `~/.cargo-make/config.toml` should be `~/.config/cargo-make.toml` or `~/.config/cargo-make/config.toml` on a typical system, but allow `$XDG_CONFIG_HOME` to override that.
* `~/.cargo-make/cache.toml` should be `~/.cache/cargo-make.toml` or `~/.cache/cargo-make/cache.toml` on a typical system, but allow `$XDG_CACHE_HOME` to override that.

**Comment by sagiegurari** [comment]
Never heard of XDG_CONFIG_HOME before and while i'm actually for it, it will cause a backward break for existing users (ya, i could update the major version but still...).
you could today also play around with the location via CARGO_MAKE_HOME env var but it will not really provide full solution to the split you described.

maybe the "safest" way would be to first honor the new directory location and if not defined there, to use the current (old) directory location.
makes sense?

I am assuming this is linux only right?

**Comment by sagiegurari** [comment]
and by the way, thanks for all the great ideas!!!

**Comment by ssokolow** [comment]
> I am assuming this is linux only right?

It's for every platform except Windows and MacOS, in the same way that pretty much everything except Windows, MacOS, and UEFI have *de facto* standardized on ELF-format binaries.

(so it applies equally to desktops based on Linux, FreeBSD, OpenBSD, NetBSD, OpenSolaris, etc.)

**Comment by sagiegurari** [comment]
implemented in the 0.12.0 branch, now running tests but if you can also verify it would be great.
i'm planning to release version 0.12 soon after I finish to verify this last issue in the 0.12 milestone.

**Comment by ssokolow** [comment]
`/home/ssokolow/.config/.cargo-make/` should really be `/home/ssokolow/.config/cargo-make/` and likewise with `.cache`.

It's very counter-intuitive and non-standard to create a hidden file/folder within the XDG directories when they're already responsible for being hidden.

I actually thought that `cargo make --help` on the new build had deleted my `config.toml` because `~/.config/car<Tab>` was turning up no completions. I tried every combination of `.config`, `.cache`, `.local/share/`, `car<Tab>` and `c*.toml<Tab>` before I decided to test to see if the `search_project_root = true` was still active and it spat out a message about moving `cache.toml` which clued me into what was going on.

**Comment by ssokolow** [comment]
Also, probably a good idea to have your migration code remove `~/.cargo-make` if it's left empty after the migration process finishes.

**Comment by sagiegurari** [comment]
valid points 👍 

just be aware that i'm not going to migrate /home/ssokolow/.config/.cargo-make to /home/ssokolow/.config/cargo-make as it is not official build yet, so you will have to rename that one time.

**Comment by sagiegurari** [comment]
0.12.0 has now been released.

**Comment by sagiegurari** [comment]
closing as it is now officially released. if there are any issues with this please reopen it.
